frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -701,39 +702,58 @@ export default ({
         // so it's good to initiate using input parameter chatRoomId
-        sbp('gi.db/archive/save', this.getArchiveKeyFromChatRoomId(chatRoomId), JSON.stringify(this.latestEvents))
+        sbp('gi.db/archive/save', this.archiveKeyFromChatRoomId(chatRoomId), JSON.stringify(this.latestEvents))
       }
     },
-    getArchiveKeyFromChatRoomId (chatRoomId) {
+    archiveKeyFromChatRoomId (chatRoomId) {
       const curChatRoomId = chatRoomId || this.currentChatRoomId
-      return `messages/${curChatRoomId}`
+      return `messages/${this.ourUsername}/${curChatRoomId}`
     },
-    refreshContent: debounce(function (from, to) {
+    refreshContent: debounce(function (to, from) {
       // NOTE: using debounce we can skip unnecessary rendering contents
       this.archiveMessageState(from)
+      this.setMessageEventListener({ to, from })
       this.setInitMessages()
-      this.setMessageEventListener({ from, to })
     }, 250)
   },
   watch: {
-    'currentChatRoomId' (to, from) {
-      if (from) {
-        this.ephemeral.messagesInitiated = false
-      }
-      this.refreshContent(from, to)
-    },
-    'summary.isJoined' (to, from) {
-      if (to) {
-        sbp('okTurtles.events/once', CONTRACT_IS_SYNCING, (contractID, isSyncing) => {
-          if (contractID === this.currentChatRoomId && isSyncing === false) {
-            this.setMessageEventListener({ from: contractID, to: contractID })
+    'summary' (to, from) {
+      const toChatRoomId = to.chatRoomId
+      const fromChatRoomId = from.chatRoomId
+      const toIsJoined = to.isJoined
+      const fromIsJoined = from.isJoined
+
+      const initAfterSynced = (chatRoomId) => {
+        const initMessagesAfterSynced = (contractID, isSyncing) => {
+          if (contractID === chatRoomId && isSyncing === false) {
             this.setInitMessages()
+            sbp('okTurtles.events/off', CONTRACT_IS_SYNCING, initMessagesAfterSynced)
           }
-        })
+        }
+
+        sbp('okTurtles.events/off', CONTRACT_IS_SYNCING, initMessagesAfterSynced)
+        sbp('okTurtles.events/on', CONTRACT_IS_SYNCING, initMessagesAfterSynced)
+      }
+
+      if (toChatRoomId !== fromChatRoomId) {
+        this.ephemeral.messagesInitiated = false
+        if (sbp('chelonia/contract/isSyncing', toChatRoomId)) {
+          this.archiveMessageState(fromChatRoomId)
+          this.setMessageEventListener({ to: toChatRoomId, from: fromChatRoomId })
+          toIsJoined && initAfterSynced(toChatRoomId)
+        } else {
+          this.refreshContent(toChatRoomId, fromChatRoomId)
+        }
+      } else if (toIsJoined && toIsJoined !== fromIsJoined) {
+        if (sbp('chelonia/contract/isSyncing', toChatRoomId)) {
+          // NOTE: toChatRoomId equals to fromChatRoomId here
+          this.setMessageEventListener({ to: toChatRoomId, from: fromChatRoomId })
+          initAfterSynced(toChatRoomId)
+        }
```

Comment:
And what if `isSyncing` returns `false` here?

Reply:
This is when user joins the channel by himself clicking the `Join Channel` button.
So the `join` action would be the last action and it couldn't return `false` here.

Reply:
This is when the user joins the channel by himself clicking the `Join channel` button, so the `join` action would be the last action of the chatroom contract.

I tested to see the `isSyncing` value, and it returned `true` at this time. I guess that's because chatroom contract is state is changed before the contract is finished to be synced. However, we should take care of the `false` value since the `join` action is the last action and the syncing would be immediately finished.